### PR TITLE
Simplify stack-to-stack moves

### PR DESCRIPTION
**Pull request by jameysharp**
Among other things, this should remove the need for the redundant-move eliminator to clean up after parallel moves.

I want to look over control flow and comments again before actually merging this but I'm stopping for the weekend. This has now passed about 7 million fuzzing iterations, and the fuzzer caught quite a few mistakes I made along the way.

### CODE REVIEW

File: src/moves.rs [hunk 1]
```diff
@@ -352,87 +337,117 @@ where
...
+            );
+            (reg, Some(save))
+        };
+
+        // Mutually exclusive flags for whether either scratch_reg or
+        // save_slot need to be restored from the other. Initially,
+        // scratch_reg has a value we should preserve and save_slot
+        // has garbage.
+        let mut scratch_dirty = false;
+        let mut save_dirty = true;
```
Comment (cfallin) [inline src/moves.rs:388]: Could we define an enum to track the state of the two locations and/or the saveslot? I think the logic here is correct but something is unsatisfying and slightly hard to follow/verify about the ad-hoc ifs -- it feels like maybe it should be a state machine.

**Review by cfallin — COMMENTED** [review]
Thanks! It seems to me that this is indeed an improvement; thanks for making a pass over the implementation.

My main comment is that it'd be nice to maybe make the final refactor push and turn the scratch/stackslot state tracking into a state machine, then document the cases. That would be a bit easier for me to reason about than the two-bools-and-an-Option controlling the cases now. (The original code also was guilty of this, so it's not like you made anything worse! But while we're here...) Thanks!

**Comment by cfallin** [comment]
(Ah, and I just saw the review request for Trevor as this tab was stale, sorry -- feel free to take over if you'd prefer otherwise I'm happy to review followup!)

**Review by elliottt — COMMENTED** [review]
Looking good! Just one question about the mutual exclusivity of the variables in the loop over moves.

File: src/moves.rs [hunk 2]
```diff
@@ -352,87 +337,117 @@ where
...
+                // make sure that the scratch register is in the correct
+                // state if this move interacts with that register.
+                if src == scratch_reg && scratch_dirty {
+                    // We're copying from the scratch register so if it
+                    // was stolen for a stack-to-stack move then we need
+                    // to make sure it has the correct contents, not
+                    // whatever was temporarily copied into it.
+                    debug_assert!(!save_dirty);
+                    result.push((save_slot.unwrap(), scratch_reg, T::default()));
+                    scratch_dirty = false;
```
[the remark this answers is not in the document]
Comment (jameysharp) [reply src/moves.rs:430]: No, `save_dirty` is false coming into this case (as checked by the debug assert) and copying its value back to the scratch register doesn't invalidate the copy in the save-slot, so `save_dirty` should still be false at the end of this case.

The requirement is that both bools can't be _true_ at the same time; it's okay if they're both false.

**Comment by jameysharp** [comment]
Yeah, I thought I'd give you a break, but your review is welcome anyway Chris. :laughing:

I felt like the two bools were the most clear way to express their state: 1) do we need to restore the scratch register after trashing it? and 2) do we need to save the scratch register's contents to the save slot? It might be nice to encode that they must not both be dirty at the same time, which would mean we've lost the value that was supposed to be in the scratch register and can't recover. I'm not completely sold that it's worth defining an enum for that, but I could be convinced if you feel strongly.

I do find it a little unsatisfying that the presence or absence of the `save_slot` only guards one of the blocks. I've convinced myself (and fuzzing confirms) that the other if statements can only be entered if that first block runs, but that's not obvious at a glance.

So I've just pushed a little rework to make it more clear how `save_slot` and `scratch_dirty` are related. Does that help?

---

Other thoughts: It would be nice to not require the caller to specify a "victim" register, partly for convenience and partly because I don't want to think about what to rename that to since I don't like the current terminology. So when `find_free_reg` fails it would be nice to pick a register from those already used by the moves.

However we can't require that there are any registers used in moves. It could happen that all the current register contents need to be preserved in-place, so there are no free registers but also no register moves. So the caller needs to provide a last-resort register in any case. Sigh.

There is potentially some room for reducing the number of generated moves though. For one thing, if we sort the moves so that moves into registers usually come after moves to the stack (except when dependencies prevent that), then we improve the odds that our selected scratch register will be overwritten after the last stack-to-stack move and we won't have to restore it from the save-slot.

If the selected register is not used as a source between the first stack-to-stack move and being used as a destination, we don't actually need the save-slot at all, because the needed values are all live in other places when we need them. Finding a register in the set of moves that satisfies this property would be a nice optimization, though I don't know how often it would come up.

A way to think about this is one could make the optimal choice in linear time by simulating the stack-to-stack move-resolution algorithm for each choice of register, counting the number of moves that would be inserted, and picking the register with the minimum count. No, I'm not proposing that we actually do that, but it does mean you can fairly easily evaluate ideas for heuristics by simulating the algorithm in your head.

File: src/moves.rs [hunk 3]
```diff
@@ -352,87 +337,117 @@ where
...
+
+            moves.with_scratch(scratch)
+        } else {
+            moves.without_scratch().unwrap()
+        };
+
+        // Do we have any stack-to-stack moves? Fast return if not.
+        let stack_to_stack = moves
+            .iter()
+            .any(|&(src, dst, _)| (self.is_stack_alloc)(src) && (self.is_stack_alloc)(dst));
```
[the remark this answers is not in the document]
Comment (jameysharp) [reply src/moves.rs]: I'm not sure it makes a big difference but sure, it's a nice change so I've gone ahead and done this.

**Comment by cfallin** [comment]
Cool, yeah, I don't feel super-strongly about reworking into a formal state machine; this is clear enough, thanks!

Re: finding a scratch register (and I agree, I didn't think hard enough about the current naming, sorry; let's call it "borrowed scratch register" or somesuch maybe?) -- I agree in principle it'd be nice to not require the caller to make the arbitrary choice; allowing the resolver the freedom to use more complex heuristics as you say. It strikes me that at this point we're several niche conditions into a rabbit hole though, and I might suggest seeing how often the case in question occurs before trying to optimize further. At least the thinking when originally building the current solution was: (i) not requiring a dedicated scratch in the `MachineEnv` was a potentially significant impact, because register pressure; but (ii) everything that happens when we hit the last-resort logic is a rare case that just needs correctness, not performance. Happy to change my mind with data of course!

File: src/moves.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (elliottt) [reply src/moves.rs:435]: `:face-palm:` thanks! From the comment at their definition, I had been assuming that they must always be mutually exclusive, but from the implementation it seems like that contract is perhaps just on the loop iteration boundary?

**Comment by jameysharp** [comment]
"Borrowed scratch register" turns out to fit nicely, done!

To be clear, I wasn't suggesting we optimize this any further now; I just wanted to note down my thoughts from making sure I understood how this works. Data indicating whether this has any impact would absolutely be good to have; I think the starting point is to look for the frequency of trace log messages saying "scratch resolver: stack-to-stack borrowing" (or prior to this PR, "scratch resolver: stack-to-stack using victim"). I would have guessed that running out of free registers is pretty common though; is it that stack-to-stack moves are rare?

**Comment by cfallin** [comment]
That's my intuition, yeah, though to be honest it's at best anecdata; I never did a detailed study. I'm satisfied of course that this PR maintains the status quo so there's no immediate need to collect better data but it would be interesting!

**Review by elliottt — APPROVED** [review]
